pageserver/src/relish_storage/storage_uploader.rs

```diff
@@ -0,0 +1,115 @@
+use std::{
+    collections::VecDeque,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+    thread,
+};
+
+use zenith_utils::zid::{ZTenantId, ZTimelineId};
+
+use crate::{relish_storage::RelishStorage, RelishStorageConfig};
+
+use super::{GenericRelishStorage, RelishStoragePath};
+
+pub struct QueueBasedRelishUploader {
+    upload_queue: Arc<Mutex<VecDeque<(ZTimelineId, PathBuf)>>>,
+}
+
+impl QueueBasedRelishUploader {
+    pub fn new(tenant_id: ZTenantId, config: &RelishStorageConfig) -> anyhow::Result<Self> {
+        let upload_queue = Arc::new(Mutex::new(VecDeque::new()));
+        let relish_storage = match config {
+            RelishStorageConfig::LocalFs(root) => {
+                GenericRelishStorage::std_local_fs(root.join(tenant_id.to_string()))?
+            }
+            RelishStorageConfig::AwsS3(s3_config) => {
+                GenericRelishStorage::rust_s3(s3_config, Some(tenant_id))?
+            }
+        };
+
+        let _handle = create_upload_thread(Arc::clone(&upload_queue), relish_storage)?;
+        Ok(Self { upload_queue })
+    }
+
+    pub fn schedule_upload(&self, timeline_id: ZTimelineId, relish_path: PathBuf) {
+        self.upload_queue
+            .lock()
+            .unwrap()
+            .push_back((timeline_id, relish_path))
+    }
+}
+
+fn create_upload_thread(
+    upload_queue: Arc<Mutex<VecDeque<(ZTimelineId, PathBuf)>>>,
+    relish_storage: GenericRelishStorage,
+) -> std::io::Result<thread::JoinHandle<()>> {
+    thread::Builder::new()
+        .name("Queue based relish uploader".to_string())
+        .spawn(move || loop {
+            upload_loop_step(&upload_queue, &relish_storage);
+        })
+}
+
+fn upload_loop_step(
+    upload_queue: &Mutex<VecDeque<(ZTimelineId, PathBuf)>>,
+    relish_storage: &GenericRelishStorage,
+) {
+    let access_queue = || upload_queue.lock().unwrap();
+
+    let mut queue_accessor = access_queue();
+    log::debug!("current upload queue length: {}", queue_accessor.len());
+    let next_upload = queue_accessor.pop_front();
```

Comment:
Taking queue items to upload one by one is not very nice and definitely deserves a better code, but feels like this code will be coupled with the future download operations (we would need to keep track of both uploaded and downloaded items), so currently just left as is for POC.

Reply:
Similarly, there can be priorities based on failure counts for any erroneous relishes; better (increasing?) timeouts when the queue is empty, etc. but that all feels so related to the choice above, so might wait a bit more before the next PR.